mava/networks/retention.py

```diff
@@ -0,0 +1,324 @@
+        first_dones = jnp.where(
+            ~jnp.any(timestep_dones, axis=1, keepdims=True),
+            jnp.full((B, 1), T),
+            jnp.argmax(timestep_dones, axis=1, keepdims=True),
+        )
+
+        # Initialize the decaying matrix 'xi'
+        xi = jnp.zeros((B, T, 1))
+        # Fill 'xi' with decaying values up until the first done step
+        for i in range(T):
+            before_first_done = i < first_dones
+            xi_i = (self.decay_kappa ** (i + 1)) * before_first_done
+            xi = xi.at[:, i, :].set(xi_i)
+
+        # Repeat the decay matrix 'xi' for all agents
+        xi = jnp.repeat(xi, self.n_agents, axis=1)
+
+        return xi
+
+
+class MultiScaleRetention(nn.Module):
+    """Multi-scale retention mechanism for Sable."""
+
+    embed_dim: int
+    n_head: int
+    n_agents: int
+    memory_config: DictConfig
+    full_self_retention: bool = False
+    decay_scaling_factor: float = 1.0
+
+    def setup(self) -> None:
+        assert self.embed_dim % self.n_head == 0, "embed_dim must be divisible by n_head"
+        # Head size
+        self.head_size = self.embed_dim // self.n_head
+
+        # Decay kappa for each head
+        self.decay_kappas = 1 - jnp.exp(
+            jnp.linspace(jnp.log(1 / 32), jnp.log(1 / 512), self.n_head)
+        )
+        self.decay_kappas = self.decay_kappas * self.decay_scaling_factor
+
+        # Initialize the weights and group norm
+        self.w_g = self.param(
+            "w_g",
+            nn.initializers.normal(stddev=1 / self.embed_dim),
+            (self.embed_dim, self.head_size),
+        )
+        self.w_o = self.param(
+            "w_o",
+            nn.initializers.normal(stddev=1 / self.embed_dim),
+            (self.head_size, self.embed_dim),
+        )
+        self.group_norm = nn.GroupNorm(num_groups=self.n_head)
+
+        # Initialize the retention mechanisms
+        self.retentions = [
+            SimpleRetention(
+                self.embed_dim,
+                self.head_size,
+                self.n_agents,
+                self.full_self_retention,
+                decay_kappa,
+                self.memory_config,
+            )
+            for decay_kappa in self.decay_kappas
+        ]
+
+        # Create an instance of the positional encoding
+        self.pe = PositionalEncoding(self.memory_config, self.embed_dim)
+
+    def __call__(
+        self,
+        key: Array,
+        query: Array,
+        value: Array,
+        hstate: Array,
+        dones: Array,
+        step_count: Array,
+    ) -> Tuple[Array, Array]:
+        """Chunkwise (default) representation of the multi-scale retention mechanism"""
+        B, C, _ = value.shape
+
+        # Positional encoding of the current step
+        key, query, value = self.pe(key, query, value, step_count)
+
+        # Per head retention
+        ret_output = jnp.zeros((B, C, self.head_size), dtype=value.dtype)
+        hs = jnp.copy(hstate)
```

Comment:
Did we ever look at just setting the new hidden states directly instead of having to make a copy?

Reply:
maybe we can do zero like?